util/deephash/deephash.go

```diff
@@ -147,8 +147,10 @@ func (h *hasher) print(v reflect.Value) (acyclic bool) {
 		// Use AppendTo methods, if available and cheap.
 		if v.CanAddr() && v.Type().Implements(appenderToType) {
 			a := v.Addr().Interface().(appenderTo)
-			scratch := a.AppendTo(h.scratch[:0])
-			w.Write(scratch)
+			size := h.scratch[:8]
+			record := a.AppendTo(size)
+			binary.LittleEndian.PutUint64(record, uint64(len(record)-len(size)))
```

Comment:
You don't really need to subtract len(size) here, because all that matters is that all lengths are distinct.

At which point you could simplify a little bit by eliminating the `size` local variable, which I think is a net readability win because it reduces the temptation to think that we should be writing the size to `size`.

```go
record := a.AppendTo(h.scratch[:8]) // save space at the beginning of record
binary.LittleEndian.PutUint64(record, uint64(len(record)))
w.Write(record)
```